mathics/eval/arithmetic.py

```diff
@@ -0,0 +1,812 @@
+            return p_eval
+        if isinstance(q_eval, Integer):
+            if isinstance(p_eval, Integer):
+                return Rational(p_eval.value, q_eval.value)
+
+        if p_eval is None:
+            p_eval = Expression(SymbolPower, p, exp)
+
+        if q_eval is None:
+            q_eval = Expression(SymbolPower, q, exp)
+        return Expression(
+            SymbolTimes, p_eval, Expression(SymbolPower, q_eval, IntegerM1)
+        )
+    # Pure imaginary base case
+    elif isinstance(base, Complex) and base.real.is_zero:
+        base = base.imag
+        if base.value < 0:
+            base = eval_negate_number(base)
+            phase = Expression(
+                SymbolPower,
+                IntegerM1,
+                eval_multiply_numbers([IntegerM1, RationalOneHalf, exp]),
+            )
+        else:
+            phase = Expression(
+                SymbolPower, IntegerM1, eval_multiply_numbers([RationalOneHalf, exp])
+            )
+        real_factor = eval_power_number(base, exp)
+
+        if real_factor is None:
+            return None
+        return Expression(SymbolTimes, real_factor, phase)
+
+    # Generic case
+    return eval_power_sympy()
+
+
+def eval_power_inexact(base: Number, exp: Number) -> BaseElement:
+    """
+    Eval base^exp for `base` and `exp` inexact numbers
+    """
+    # If both base and exponent are exact quantities,
+    # use sympy.
+    prec = min_prec(base, exp)
+    if prec is not None:
+        is_machine_precision = base.is_machine_precision() or exp.is_machine_precision()
+        if is_machine_precision:
+            number = mpmath.power(base.to_mpmath(), exp.to_mpmath())
+            return from_mpmath(number)
+        else:
+            with mpmath.workprec(prec):
+                number = mpmath.power(base.to_mpmath(), exp.to_mpmath())
+                return from_mpmath(number, dps(prec))
+
+
+def eval_sign(expr: BaseElement) -> Optional[BaseElement]:
+    """
+    evaluates Sign[expr]
+    """
+    if isinstance(expr, Atom):
+        return eval_sign_number(expr)
+    if expr.has_form("Power", 2):
+        base, exp = expr.elements
+        if exp.is_zero:
+            return Integer1
+        if isinstance(exp, (Integer, Real, Rational)):
+            sign = eval_sign(base) or Expression(SymbolSign, base)
+            return Expression(SymbolPower, sign, exp)
+        if isinstance(exp, Complex):
+            sign = eval_sign(base) or Expression(SymbolSign, base)
+            return Expression(SymbolPower, sign, exp.real)
+        if test_arithmetic_expr(exp):
+            sign = eval_sign(base) or Expression(SymbolSign, base)
+            return Expression(SymbolPower, sign, exp)
+        return None
+    if expr.has_form("Exp", 1):
+        exp = expr.elements[0]
+        if isinstance(exp, (Integer, Real, Rational)):
+            return Integer1
+        if isinstance(exp, Complex):
+            return Expression(SymbolExp, exp.imag)
```

Comment:
My editor and flake8 tell me that SymbolExp is not defined.

Reply:
> With respect to
> 
> > conversions back and forward to SymPy.
> 
> I would like more details here too. Obviously we want to reduce this, but generally the ideal flow is convert into SymPy as much as possible and convert back to Mathics as late or as little as possible. In fact this idea is true for other things as well, like numpy arrays, constant literals that map into Python, etc.

Suppose you are going to evaluate

```
3^(1/2) +  5 * 3^(-1/2) + 3 * Pi^2 
```
`1/2` is converted to sympy twice back and forward, to see that the expression didn't changed, so `1/2` is kept. Then  we have the evaluation of `3^(-1/2)` and `3^(-1/2)` where `3` and `1/2`, and `3^(1/2)` are converted again back and forward, to again get the same result. And again in the multiplication by 5, and again in the addition, to get at the end the same expression.

Reply:
All of this can be addressed by having a special form of Expression that is known to have a SymPy equivalent and that is saved.  We will probably do the same for literals that have exact Python correspondences, and numerous other things. 

If this is the only thing that being addressed, please don't do it this way.

Reply:
The problem is that for certain expressions, the replacement rules in WMA and SymPy are different, which makes that new  rules written in WL fails to be applied, like in the case of DirectedInfinity. Actually, I did all of this because it was not possible to fix these rules without disentangling all the arithmetic module.

Reply:
I think here we need more knowledgeable conversion routines between the Mathics and SymPy representations.

Reply:
> Guder

That part of the PR are in this branch: https://github.com/Mathics3/mathics-core/tree/tests_from_arithmetic_refactor

However, the new tests do not pass without the other changes.

Reply:
A number of them do, so those could co in. The ones that don't can be commented out to remind us to address various issues.

Reply:
@mmatera if you would like me to make a stab and extracting the non-controversial parts, I'll do.